artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java

```diff
@@ -599,6 +600,29 @@ private long checkMinPage(Collection<PageSubscription> cursorList) {
 
    }
 
+   private void deliverIfNecessary(Collection<PageSubscription> cursorList) {
+      long minPage = Long.MAX_VALUE;
+      PageSubscription slowSubscription = null;
+      int nonEmptyCursorNum = 0;
+
+      for (PageSubscription cursor : cursorList) {
+         long firstPage = cursor.getFirstPage();
+
+         // the cursor will return -1 if the cursor is empty
+         if (firstPage >= 0) {
+            nonEmptyCursorNum++;
+            if (firstPage < minPage) {
+               minPage = firstPage;
+               slowSubscription = cursor.getQueue().getMessageCount() == 0 ? cursor : null;
+            }
+         }
+      }
+
+      if (nonEmptyCursorNum > 1 && slowSubscription != null) {
```

Comment:
I was talking to @franz1981  here where I am.. couldn't we just call deliverAsync on every queue after cleanup?

Reply:
this is saving a deliveryAsync call.. but I"m just wondering if it would be too costly to do the check on every queue?

Reply:
@clebertsuconic @wy96f 
Not sure but deliverIfNecessary could be written simlar to this one?
```
   private void deliverIfNecessary(Collection<PageSubscription> cursorList, long minPage) {
      for (PageSubscription cursor : cursorList) {
         long firstPage = cursor.getFirstPage();
         if (firstPage == minPage) {
            if (cursor.getQueue().getMessageCount() == 0) {
               cursor.getQueue().deliverAsync();
               break;
            }
         }
      }
   }
```
As long as minPage is not Long.MAX_VALUE (it shouldn't be)...

Reply:
> this is saving a deliveryAsync call.. but I"m just wondering if it would be too costly to do the check on every queue?

All PageSubcriptions use the same executor and it will do many jobs such as PageSubscriptionImpl::cleanupEntries, DepageRunner, DeliverRunner, ExpiryScanner, PageCursorProviderImpl::scheduleCleanup, PageCursorProviderImpl::cleanup, etc. We'd better prevent these  extra needless deliverAsync calls.

Reply:
@wy96f nice! So please change the code with the refined version of `deliverIfNecessary` and squash the commits into one :+1: